### ISSUE TITLE
Dashboard server seems unable to keep up with new log messages

### ISSUE THREAD

**Issue by JustAnotherArchivist** [ArchiveTeam/ArchiveBot#333]
It appears that we've reached a point where the dashboard server is unable to keep up with the log messages. The dashboard is noticeably laggy, i.e. log lines do not appear in real time; I haven't measured the delay though. The `dashboard/app.rb` process seems to constantly use one CPU core of the control node.

**Comment by yipdw** [comment]
That lasted longer than I expected :P 

The good news is that the dashboard webapp isn't doing that much; it's taking messages, doing some reformatting, and spewing them out on a websocket.  Maybe a good project to reimplement in a way that's faster and doesn't leak memory everywhere?

**Comment by JustAnotherArchivist** [comment]
After it got much worse over the past few days (2-minute lag), I wrote a replacement dashboard WebSocket server in Python this weekend, and I'm happy to report that it's running fine! Uses about 10 MB RSS baseline plus 1-2 MB per client and <20 % CPU nearly always (near-zero if no clients are connected). The memory usage is still going up slowly, but that might just be Python not running the GC very aggressively. It's far from problematic though; after about 26 hours of runtime, RSS is at 72 MiB.

An early version of this replacement server pretty much confirmed what I was suspecting already before: it's not a memory leak but a lack of backpressure handling. Backpressure in this context means that a client doesn't read from the WebSocket fast enough due to e.g. a slow connection, a potato CPU, or a broken client. Because the server keeps trying to send more messages, this builds up a large buffer in memory of messages that haven't been sent out to a particular client yet. I *assume* that the Ruby server didn't handle this, but I didn't feel like digging into Reel to confirm this. I *know* that the initial version of my Python server didn't handle it; I intentionally did that to confirm this issue. Due to how I structured the code, adding the necessary handling (dropping messages when the backlog becomes too large) was very simple though, and the memory "leak" vanished after that.

The code will be in a PR in the next few days.
